File: src/lsp-client.js
import net from 'net';
import { EventEmitter } from 'events';

/**
 * Language Server Protocol client for Godot
 * 
 * LSP uses JSON-RPC 2.0 over TCP with Content-Length headers (similar to DAP)
 */
export class LSPClient extends EventEmitter {
  constructor(host = '127.0.0.1', port = 6005) {
    super();
    this.host = host;
    this.port = port;
    this.socket = null;
    this.buffer = '';
    this.pendingRequests = new Map();
    this.sequence = 1;
    this.connected = false;
    this.initialized = false;
  }

  /**
   * Connect to the Godot LSP port
   */
  async connect() {
    return new Promise((resolve, reject) => {
      this.socket = net.createConnection(this.port, this.host, () => {
        this.connected = true;
        console.error(`Connected to Godot LSP port at ${this.host}:${this.port}`);
        
        // Send initialize request
        this.sendRequest('initialize', {
          processId: process.pid,
          clientInfo: {
            name: 'godot-mcp-server',
            version: '1.0.0'
          },
          rootUri: null,
          capabilities: {
            textDocument: {
              hover: {
                contentFormat: ['markdown', 'plaintext']
              },
              completion: {
                completionItem: {
                  snippetSupport: true,
                  documentationFormat: ['markdown', 'plaintext']
                }
              },
              definition: {
                linkSupport: true
              },
              documentSymbol: {
                hierarchicalDocumentSymbolSupport: true
              }
            },
            workspace: {
              symbol: {
                symbolKind: {
                  valueSet: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]
                }
              }
            }
          }
        }).then((result) => {
          this.initialized = true;
          // Send initialized notification
          this.sendNotification('initialized', {});
          console.error('LSP initialized successfully');
          resolve(result);
        }).catch(reject);
      });

      this.socket.on('data', (data) => {
        this.handleData(data);
      });

      this.socket.on('error', (error) => {
        console.error('LSP socket error:', error.message);
        this.connected = false;
        reject(error);
      });

      this.socket.on('close', () => {
        console.error('LSP connection closed');
        this.connected = false;
        this.initialized = false;
      });

      // Set a timeout for connection
      setTimeout(() => {
        if (!this.connected) {
          reject(new Error('LSP connection timeout'));
        }
      }, 5000);
    });
  }

  /**
   * Handle incoming data from the socket
   */
  handleData(data) {
    this.buffer += data.toString();
    
    // Process complete messages
    while (true) {
      const match = this.buffer.match(/^Content-Length: (\d+)\r?\n\r?\n/);
      if (!match) break;
      
      const contentLength = parseInt(match[1]);
      const headerLength = match[0].length;
      const totalLength = headerLength + contentLength;
      
      if (this.buffer.length < totalLength) break;
      
      const messageContent = this.buffer.substring(headerLength, totalLength);
      this.buffer = this.buffer.substring(totalLength);
      
      try {
        const message = JSON.parse(messageContent);
        this.handleMessage(message);
      } catch (error) {
        console.error('Error parsing LSP message:', error);
      }
    }
  }

  /**
   * Handle a parsed LSP message
   */
  handleMessage(message) {
    if (message.id && this.pendingRequests.has(message.id)) {
      // Response to our request
      const handler = this.pendingRequests.get(message.id);
      this.pendingRequests.delete(message.id);
      
      if (message.error) {
        handler.reject(new Error(message.error.message || 'LSP request failed'));
      } else {
        handler.resolve(message.result);
      }
    } else if (message.method) {
      // Notification or request from server
      this.emit('notification', message);
    }
  }

  /**
   * Send an LSP request
   */
  async sendRequest(method, params = {}) {
    if (!this.connected) {
      throw new Error('Not connected to Godot LSP port');
    }

    const id = this.sequence++;
    const request = {
      jsonrpc: '2.0',
      id: id,
      method: method,
      params: params
    };

    const message = JSON.stringify(request);
    const header = `Content-Length: ${Buffer.byteLength(message)}\r\n\r\n`;
    
    return new Promise((resolve, reject) => {
      this.pendingRequests.set(id, { resolve, reject });
      
      this.socket.write(header + message, (error) => {
        if (error) {
          this.pendingRequests.delete(id);
          reject(error);
        }
      });

      // Timeout after 10 seconds
      setTimeout(() => {
        if (this.pendingRequests.has(id)) {
          this.pendingRequests.delete(id);
          reject(new Error('LSP request timeout'));
        }
      }, 10000);
    });
  }

  /**
   * Send an LSP notification (no response expected)
   */
  sendNotification(method, params = {}) {
    if (!this.connected) {
      throw new Error('Not connected to Godot LSP port');
    }

    const notification = {
      jsonrpc: '2.0',
      method: method,
      params: params
    };

    const message = JSON.stringify(notification);
    const header = `Content-Length: ${Buffer.byteLength(message)}\r\n\r\n`;
    
    this.socket.write(header + message);
  }

  /**
   * Notify LSP that a document was opened
   */
  didOpenTextDocument(uri, languageId = 'gdscript', version = 1, text = '') {
    this.sendNotification('textDocument/didOpen', {
      textDocument: {
        uri: uri,
        languageId: languageId,
        version: version,
        text: text
      }
    });
  }

  /**
   * Get diagnostics for a document
   * Note: LSP sends diagnostics as notifications, so we need to listen for them
   */
  async getDocumentDiagnostics(uri) {
    // In LSP, diagnostics are sent as notifications after didOpen or didChange
    // We'll need to track them separately
    throw new Error('Use diagnostic notifications from the server instead');
  }

  /**
   * Get hover information at a position
   */
  async hover(uri, line, character) {
    if (!this.initialized) {
      throw new Error('LSP not initialized');
    }

    return this.sendRequest('textDocument/hover', {
      textDocument: { uri: uri },
      position: { line: line, character: character }
    });
  }

  /**
   * Get definition location
   */
  async definition(uri, line, character) {
    if (!this.initialized) {
      throw new Error('LSP not initialized');
    }

    return this.sendRequest('textDocument/definition', {
      textDocument: { uri: uri },
      position: { line: line, character: character }
    });
  }

  /**
   * Get completion suggestions
   */
  async completion(uri, line, character) {
    if (!this.initialized) {
      throw new Error('LSP not initialized');
    }

    return this.sendRequest('textDocument/completion', {
      textDocument: { uri: uri },
      position: { line: line, character: character }
    });
  }

  /**
   * Get document symbols
   */
  async documentSymbols(uri) {
    if (!this.initialized) {
      throw new Error('LSP not initialized');
    }

    return this.sendRequest('textDocument/documentSymbol', {
      textDocument: { uri: uri }
    });
  }

  /**
   * Search for workspace symbols
   */
  async workspaceSymbols(query = '') {
    if (!this.initialized) {
      throw new Error('LSP not initialized');
    }

    return this.sendRequest('workspace/symbol', {
      query: query
    });
  }

  /**
   * Disconnect from the LSP port
   */
  disconnect() {
    if (this.socket) {
      if (this.initialized) {
        // Send shutdown request
        this.sendRequest('shutdown', {}).then(() => {
          this.sendNotification('exit', {});
          this.socket.destroy();
        }).catch(() => {
          this.socket.destroy();
        });
      } else {
        this.socket.destroy();
      }
      this.socket = null;
      this.connected = false;
      this.initialized = false;
    }
  }

  /**
   * Check if connected
   */
  isConnected() {
    return this.connected;
  }

  /**
   * Check if initialized
   */
  isInitialized() {
    return this.initialized;
  }
}

File: src/godot_mcp/lsp_client.py
"""Language Server Protocol client for Godot.

LSP uses JSON-RPC 2.0 over TCP with Content-Length headers (similar to DAP).
"""

import asyncio
import json
import os
import sys


def log(*args):
    print(*args, file=sys.stderr, flush=True)


class LSPClient:
    def __init__(self, host: str = "127.0.0.1", port: int = 6005):
        self.host = host
        self.port = port
        self.reader: asyncio.StreamReader | None = None
        self.writer: asyncio.StreamWriter | None = None
        self.pending_requests: dict[int, asyncio.Future] = {}
        self.sequence = 1
        self.connected = False
        self.initialized = False
        self.diagnostics_map: dict[str, list] = {}
        self._reader_task: asyncio.Task | None = None

    async def connect(self):
        try:
            self.reader, self.writer = await asyncio.wait_for(
                asyncio.open_connection(self.host, self.port), timeout=5)
        except (asyncio.TimeoutError, OSError) as e:
            raise ConnectionError(f"LSP connection failed at {self.host}:{self.port}: {e}")

        self.connected = True
        log(f"Connected to Godot LSP port at {self.host}:{self.port}")
        self._reader_task = asyncio.get_event_loop().create_task(self._read_loop())

        result = await self.send_request("initialize", {
            "processId": os.getpid(),
            "clientInfo": {"name": "godot-mcp-server", "version": "1.0.0"},
            "rootUri": None,
            "capabilities": {
                "textDocument": {
                    "hover": {"contentFormat": ["markdown", "plaintext"]},
                    "completion": {
                        "completionItem": {
                            "snippetSupport": True,
                            "documentationFormat": ["markdown", "plaintext"],
                        }
                    },
                    "definition": {"linkSupport": True},
                    "documentSymbol": {"hierarchicalDocumentSymbolSupport": True},
                    "publishDiagnostics": {"relatedInformation": True},
                },
                "workspace": {
                    "symbol": {"symbolKind": {"valueSet": list(range(1, 27))}}
                },
            },
        })
        self.initialized = True
        await self.send_notification("initialized", {})
        log("LSP initialized successfully")
        return result

    async def _read_loop(self):
        buffer = b""
        try:
            while True:
                data = await self.reader.read(65536)
                if not data:
                    break
                buffer += data
                while True:
                    header_end = buffer.find(b"\r\n\r\n")
                    if header_end == -1:
                        break
                    header = buffer[:header_end].decode("utf-8", "replace")
                    content_length = None
                    for line in header.split("\r\n"):
                        if line.lower().startswith("content-length:"):
                            content_length = int(line.split(":", 1)[1].strip())
                    if content_length is None:
                        buffer = buffer[header_end + 4:]
                        continue
                    total = header_end + 4 + content_length
                    if len(buffer) < total:
                        break
                    body = buffer[header_end + 4:total]
                    buffer = buffer[total:]
                    try:
                        self._handle_message(json.loads(body))
                    except json.JSONDecodeError as e:
                        log(f"Error parsing LSP message: {e}")
        except Exception as e:
            log(f"LSP socket error: {e}")
        finally:
            self.connected = False
            self.initialized = False
            log("LSP connection closed")

    def _handle_message(self, message: dict):
        msg_id = message.get("id")
        if msg_id is not None and msg_id in self.pending_requests:
            future = self.pending_requests.pop(msg_id)
            if not future.done():
                if message.get("error"):
                    future.set_exception(
                        RuntimeError(message["error"].get("message") or "LSP request failed"))
                else:
                    future.set_result(message.get("result"))
        elif message.get("method"):
            # Notification
            if message["method"] == "textDocument/publishDiagnostics":
                params = message.get("params") or {}
                if params.get("uri"):
                    self.diagnostics_map[params["uri"]] = params.get("diagnostics", [])

    async def send_request(self, method: str, params: dict | None = None, timeout: float = 10):
        if not self.connected:
            raise ConnectionError("Not connected to Godot LSP port")

        msg_id = self.sequence
        self.sequence += 1
        body = json.dumps({
            "jsonrpc": "2.0", "id": msg_id, "method": method, "params": params or {}
        }).encode("utf-8")

        future = asyncio.get_event_loop().create_future()
        self.pending_requests[msg_id] = future

        self.writer.write(f"Content-Length: {len(body)}\r\n\r\n".encode() + body)
        await self.writer.drain()

        try:
            return await asyncio.wait_for(future, timeout=timeout)
        except asyncio.TimeoutError:
            self.pending_requests.pop(msg_id, None)
            raise TimeoutError("LSP request timeout")

    async def send_notification(self, method: str, params: dict | None = None):
        if not self.connected:
            raise ConnectionError("Not connected to Godot LSP port")
        body = json.dumps({
            "jsonrpc": "2.0", "method": method, "params": params or {}
        }).encode("utf-8")
        self.writer.write(f"Content-Length: {len(body)}\r\n\r\n".encode() + body)
        await self.writer.drain()

    async def did_open_text_document(self, uri: str, language_id: str = "gdscript",
                                     version: int = 1, text: str = ""):
        await self.send_notification("textDocument/didOpen", {
            "textDocument": {
                "uri": uri, "languageId": language_id, "version": version, "text": text
            }
        })

    async def get_document_diagnostics(self, uri: str | None = None):
        """LSP sends diagnostics as notifications, so return cached values."""
        if uri:
            return self.diagnostics_map.get(uri, [])
        return dict(self.diagnostics_map)

    def _require_initialized(self):
        if not self.initialized:
            raise RuntimeError("LSP not initialized")

    async def hover(self, uri: str, line: int, character: int):
        self._require_initialized()
        return await self.send_request("textDocument/hover", {
            "textDocument": {"uri": uri},
            "position": {"line": line, "character": character},
        })

    async def definition(self, uri: str, line: int, character: int):
        self._require_initialized()
        return await self.send_request("textDocument/definition", {
            "textDocument": {"uri": uri},
            "position": {"line": line, "character": character},
        })

    async def completion(self, uri: str, line: int, character: int):
        self._require_initialized()
        return await self.send_request("textDocument/completion", {
            "textDocument": {"uri": uri},
            "position": {"line": line, "character": character},
        })

    async def document_symbols(self, uri: str):
        self._require_initialized()
        return await self.send_request("textDocument/documentSymbol", {
            "textDocument": {"uri": uri}
        })

    async def workspace_symbols(self, query: str = ""):
        self._require_initialized()
        return await self.send_request("workspace/symbol", {"query": query})

    def disconnect(self):
        if self._reader_task:
            self._reader_task.cancel()
            self._reader_task = None
        if self.writer:
            self.writer.close()
            self.writer = None
        self.reader = None
        self.connected = False
        self.initialized = False

    def is_connected(self) -> bool:
        return self.connected

    def is_initialized(self) -> bool:
        return self.initialized
